### PR TITLE
Paper: Reproducible Environments for Reproducible Results

### PR DESCRIPTION
**Pull request by werwty**
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.

### CODE REVIEW

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 1]
```diff
@@ -0,0 +1,633 @@
...
+:equal-contributor:
+
+:bibliography: mybib
+
+--------------------------------------------------
+Reproducible Environments for Reproducible Results
+--------------------------------------------------
+
+.. class:: abstract
+
```
Comment (WorklnProgress) [inline papers/zhang_macdonald/zhang_macdonald.rst:18]: I would recommend seeing the paper built by the build server and having some consistency in the font (size can differ) between abstract and the main paper.

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 2]
```diff
@@ -0,0 +1,633 @@
...
+tool for each content type. Package management is inherently complicated.
+Each content type handles the complexities in a different way- usually tools
+are built and optimized for a single content type. Context switching between
+these tools consume human RAM cycles.
+
+
+Pulp
+----
+
+Pulp is an open source repository manager[2]_ that can be used to create
```
Comment (WorklnProgress) [inline papers/zhang_macdonald/zhang_macdonald.rst]: I think this would benefit a lot from a diagram showing the workflow and what steps are repeatable etc. As it stands now, it's really difficult to get an understanding of the exact process and reasons behind the steps.

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 3]
```diff
@@ -0,0 +1,633 @@
...
+Development
+-----------
+
+1. Set up a pulp repository to mirror a subset of packages from PyPI that
+are used by your lab.
+
+Create a Repository
+
+.. code-block:: bash
+
```
Comment (WorklnProgress) [inline papers/zhang_macdonald/zhang_macdonald.rst]: Some code block formatting issues. If you see the paper on the build server, you can see that some code blocks leak to the second column.

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 4]
```diff
@@ -0,0 +1,633 @@
...
+    .. code-block:: bash
+
+        http://trypulp.org/pulp/content/
+        top-secret-development/simple/
+
+
+Summary
+=======
+
+For researches who use code in their methods, it is crucual to consider the
```
Comment (WorklnProgress) [inline papers/zhang_macdonald/zhang_macdonald.rst]: typo, crucial?

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 5]
```diff
@@ -0,0 +1,633 @@
...
+
+Measuring Reproducibility
+=========================
+
+Two factors have to be considered when we think about reproducibility.
+
+Complete reproducibility is having the researcher and reviewer share identical
+'bits' of the necessary system, program, and dependencies.
+
+Vandewalle identifies several necessities for complete reproducibility
```
Comment (WorklnProgress) [inline papers/zhang_macdonald/zhang_macdonald.rst]: may be bullets to emphasize what the two factors are clearly?

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 6]
```diff
@@ -0,0 +1,633 @@
...
+maintain backward compatibility. However, there are other tools with rich
+feature sets that simplify the process. In particular,
+pipenv_ uses hashing and virtual environments by
+default for a smooth experience.
+
+
+Ansible
+-------
+
+Ansible_ is an IT automation tool. It can configure systems, deploy software,
```
Comment (WorklnProgress) [inline papers/zhang_macdonald/zhang_macdonald.rst]: May be emphasize that there are other tools like Ansible and explain the reason behind describing only Ansible just like at the end of the paper, a section explains why `Docker` over the more general `LXC` etc has been chosen.

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 7]
```diff
@@ -0,0 +1,633 @@
...
+----------
+
+Containers_ [1]_ "are technologies that allow you to package and isolate
+applications with their entire runtime environment—all of the files
+necessary to run." Applied to the scienctific field this means that each
+container will contain an image of your system, a copy of your source code,
+installed dependencies, and data used. These are stored in a static file called
+an Image.
+
+This Image can be given to peer reviewers and other collaborators as a baseline
```
Comment (WorklnProgress) [inline papers/zhang_macdonald/zhang_macdonald.rst]: may be `image` instead of Image

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 8]
```diff
@@ -0,0 +1,633 @@
...
+to run your research. However the Image itself is opaque, and it is hard to
+tell what dependencies have been installed on the image without substantial
+inspection.  It is recommended that the Image is built from a Dockerfile for
+full transparency.
+
+A Dockerfile is a text document that contains all the commands a user could call
```
Comment (WorklnProgress) [inline papers/zhang_macdonald/zhang_macdonald.rst]: Yaml file instead of text document?

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 9]
```diff
@@ -0,0 +1,633 @@
...
+used in large production environments for about 4 years. Pulp v.2 is useful for
+owning a pipe, and versioned repositories can be implemented by the user.
+
+Pulp v.3 is currently in beta, and supports python and ansible plugins.
+Pulp v.3 natively versions repositories, which makes it ideal for careful
+management optimized for reproducibility.
+
+
+Concepts_:
+    Content units (like Python Wheel, or Ansible Role) in Pulp are organized by
```
Comment (WorklnProgress) [inline papers/zhang_macdonald/zhang_macdonald.rst]: General formatting with color, beginning of paragraphs etc in this section, can be easily seen on the build server.

**Review by WorklnProgress — COMMENTED** [review]
Overall a good draft, some formatting issues. A few suggestions for improving clarity. I think the paper would greatly benefit by clearly identifying how Pulp is a confluence of all the things mentioned above it. A suggestion to do this would be identify drawback of each technology in isolation and then in Pulp portion, add a section that addresses each. Also a diagram showing the Pulp workflow would be greatly beneficial.

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 10]
```diff
@@ -0,0 +1,633 @@
...
+and tools that can make this easier. A common solution for high level
+dependencies is to explicitly "pin" them :cite:`Wilson`. In python this would be
+running
+
+.. code-block:: bash
+
+   pip freeze > requirements.txt
+
+
+But this only manages package dependencies and not system dependencies. And,
```
Comment (dalippa) [inline papers/zhang_macdonald/zhang_macdonald.rst]: Can you give an example of system dependencies? Are you referring to configuration? OS? System architecture?

File: papers/zhang_macdonald/zhang_macdonald.rst (hunk 5 above)
Comment (dalippa) [reply papers/zhang_macdonald/zhang_macdonald.rst]: Bullets or a semicolon delimited list after a colon work well.

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 11]
```diff
@@ -0,0 +1,633 @@
...
+A good tool, and management system will balance complete reproducibility and
+code entropy.
+
+Tools
+=====
+
+Published Source Code
+---------------------
+
+Scholarly research containing descriptions of methodology is no longer
```
Comment (dalippa) [inline papers/zhang_macdonald/zhang_macdonald.rst]: This is the third time on this page that source code isn't sufficient. Consider removing this sentence.

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 12]
```diff
@@ -0,0 +1,633 @@
...
+sufficient.  For standalone scripts, publishing source code might be
+acceptable, but as computational systems grow more complex, this method becomes
+more unreliable. Nontrivial research often depends on other external
+libraries for everything from left-padding a line to building scalable machine
+learning. This "has led to an ever larger and more complex black box between
+what was actually done and what is described in literature." :cite:`Boettiger`
```
Comment (dalippa) [inline papers/zhang_macdonald/zhang_macdonald.rst]: What about hosted IPython notebooks? This has been commonplace in the Python community for many years now.

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 13]
```diff
@@ -0,0 +1,633 @@
...
+
+Publishing source code in papers makes them inflexible to change-- bugs fixed
```
Comment (dalippa) [inline papers/zhang_macdonald/zhang_macdonald.rst]: This is a good point, but can be solved by version control. Scripts that automatically check out the latest on master, or equivalent, have been part of continuous integration-continuous deployment (CI-CD) systems for a long time now. Why is it not possible to extend the same principles to such scripts?

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 14]
```diff
@@ -0,0 +1,633 @@
...
+GitHub and Other Version Control Software
+-----------------------------------------
+
+Using an online git repository is a great way to keep track of source code
+:cite:`Wilson`.  With git you can easily track changes you make to data and
+software. Git identifies commits by a unique hash, which can be used to
+reference a specific point in the source code.
+
+What git lacks is the ability to do environmental management, it is not a
+package manager. System dependencies in git can only be documented-- and need
```
Comment (dalippa) [inline papers/zhang_macdonald/zhang_macdonald.rst]: This is incorrect. Dependencies in git can be used to manage the environment via submodules (https://git-scm.com/book/en/v2/Git-Tools-Submodules).

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 15]
```diff
@@ -0,0 +1,633 @@
...
+
+The requirements file can now be used to recreate the same environment using
+the same versions.
+
+.. code-block:: bash
+
+   $ mkvirtualenv separate-env
+   (separate-env) $ pip install -r scipy-requirements.txt
+
+For Python users who need to guarantee deterministic builds, another step is
```
Comment (dalippa) [inline papers/zhang_macdonald/zhang_macdonald.rst]: For deterministic builds, it might be best to have CI-CD systems build an egg rather than leave it to the user to build the system.

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 16]
```diff
@@ -0,0 +1,633 @@
...
+    to a distribution, which determines how and where a publication is served.
+
+
+With the versioned repository, and publication, the following workflow can be used
+to manage reproducibility:
+
+Development
+-----------
+
+1. Set up a pulp repository to mirror a subset of packages from PyPI that
```
Comment (dalippa) [inline papers/zhang_macdonald/zhang_macdonald.rst]: What does pulp bring to the table that other current mechanisms do not?

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 17]
```diff
@@ -0,0 +1,633 @@
...
+.. code-block:: bash
+
+    http POST http://trypulp.org/pulp/api/v3/repositories/e81221c3-
+    9c7a-4681-a435-aa74020753f2/remotes/python/3750748b-781f-48df-
+    9734-df014b2a11b4/'sync/'
+    repository=http://trypulp.org/pulp/
+    api/v3/repositories/e81221c3-9c7a-4681-a435-aa74020753f2/
+
+
+Publish a repository
```
Comment (dalippa) [inline papers/zhang_macdonald/zhang_macdonald.rst]: Can you contrast this with git? What's the difference between publish and a push?

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 18]
```diff
@@ -0,0 +1,633 @@
...
+=======
+
+For researches who use code in their methods, it is crucual to consider the
+reproducibility of the software environments they use. Excellent research can
+become nearly impossible to replicate because of the difficulty of maintaining
+a reliable dependency chain. By using the tools best practices developed for
+software engineering, researchers can take steps to prevent code entropy and
+preserve the efficacy of their work.
+
+Acknowledgements
```
Comment (dalippa) [inline papers/zhang_macdonald/zhang_macdonald.rst]: Future work section?

**Review by dalippa — CHANGES_REQUESTED** [review]
Overall, the paper makes a weak case and is mostly a survey of existing software methods for ensuring reproducibility and a listing of API calls to use pulp. Very little of the paper focuses on the novelty of pulp, comparing and contrasting with existing systems (especially git), and notably excludes CI-CD systems and IPython notebooks that are currently top choices for reproducing results.

The structure of the paper also needs work: the paper narrative doesn't flow, does not describe the benefits of pulp (which was introduced abruptly and briefly), etc.

**Comment by cyrush** [comment]
Overall Recommendations:
- needs to be clearly framed as focused on introducing Pulp
- needs a more focused related work section
- needs more details on Pulp, how it address shortfalls with other tools, etc

Review Details:

Since this is focused on Pulp, I recommend you include Pulp in the title.

The introduction provides motivation and states it explores several different methods for managing environmental reproducibility. I don't think the narrative supports the "several different methods" aspiration. Environmental Management section only focuses on introducing Pulp.  No need to make such broad claims, I recommend removing this claim and simply focusing the overall narrative on introducing Pulp. 

This paper needs a more focused related work section. 

There are many existing efforts in this space (conda's environment support, binder, popper (https://github.com/systemslab/popper), etc) 

It would be intractable to do a deep, in-depth comparison with all of these tools, but since they are closer to Pulp than tools which are discussed (published source code, using version control), they should be included. 

The `Tools` section as written is part background, part related work, part components leveraged by Pulp -- this makes it hard to parse. 

There are also quite a few narrative gaps, here are a few examples:

The Containers subsection in Tools introduces the concept of a Container and Image, but then jumps
to talking about Docker files without introducing Docker. What is missing is a statement that Docker is a specific container technology.  The listed drawbacks of containers reference specific issues related to Docker (from Docker's bug tracker). Yes, Docker is a behemoth and often colloquially synonymous with Containers, but there are other Container technologies.

The Python Packaging section talks about how "Pip was selected", but it is not clear what it was selected for (Pulp hasn't really even been introduced at this point )

The paper needs more content related to Pulp. The Pulp subsection provides a specific example workflow using Pulp -- but it does not clearly outline Pulp's design, which parts of the Tools section are leveraged, or how it addresses listed limitations in other tools, etc.

**Comment by deniederhut** [comment]
Howdy @werwty ! How is the rewrite going?

**Comment by werwty** [comment]
Hey, it's going, I will push out the revision next Monday.

**Comment by deniederhut** [comment]
Hm... that won't leave much time for the reviewers to take another pass through the paper. Is there any chance you could push something sooner?

**Comment by werwty** [comment]
Ack, we will have the revision ready by tonight/early tomorrow morning

**Comment by dalippa** [comment]
Hi @werwty. You mentioned you would have a new revision available. Will it be up by Monday evening for re-review?

**Comment by werwty** [comment]
Revision is up and ready for re-review. Sorry about the delay.

**Comment by deniederhut** [comment]
Dear @WorklnProgress and @cyrush do you feel this paper is now ready for inclusion in the proceedings?

**Comment by cyrush** [comment]
@deniederhut  I plan to review today.

**Comment by deniederhut** [comment]
Hi @WorklnProgress ! We are compiling the names of our reviewers for acknowledgment in the proceedings. There is not a name attached to your GitHub handle - would you mind sending it my way? If you don't want it publicly listed, that's okay too.

**Comment by WorklnProgress** [comment]
Update my name in the profile. I will review once more tonight. Thank you!

**Comment by cyrush** [comment]
This reads better and addressed several requests. I still think it could use a related work section -- there are several things close or trying to solve similar problems.  The paper is still quite short (~3 pages w/o references). There isn't a hard threshold on number of pages, but it does seem more text could clarify the presentation.


There are still a few narrative jumps:  

1) Ansible Roles are referenced before they are described.

2)  "Pulp’s tasking system requires running rq.", Whats rq?


A diagram for the example workflow would help guide readers through that section.


Here are a few minor things I spotted:

Pulp Architecture Diagram:

For the diagram, the titles of the components aren't separated from their descriptions, for example:

1) WSGI application Pulp’s web application is ...
component.

Maybe add a '-'?

1) WSGI application - Pulp’s web application is ...


Containers subsection:

The sentence after Container is quoted, but not sure why + there is no attribution for the quote

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 19]
```diff
@@ -0,0 +1,368 @@
...
+
+
+Introduction
+============
+
+Reliability of research, and therefore reproducibility, is the backbone of cumulative knowledge,
+and has been identified as a bottleneck for efficient scientific progress. Open
+data, reproducibility incentives, and comprehensive experimental procedures are all key areas for
+improvement, but the subtle problem of environmental reproducibility is in need of more
+attention. Because software is an integral part of modern science, software distributions and its
```
Comment (dalippa) [inline papers/zhang_macdonald/zhang_macdonald.rst:39]: "and their"

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 20]
```diff
@@ -0,0 +1,368 @@
...
+software industry tools and workflows to scientific code, focusing on the Pulp project and the Python
+community.
+
+Software Packaging Basics
+=========================
+
+Software packaging and distribution is an inherently complex problem domain because software is
+dependent on other software. Ensuring that code can be installed with compatible requirements is a
+fundamental problem in the packaging domain. The needs of users, developers, and administrators
+frequently interfere, leading to increasingly complex solutions. Various communities (often formed
```
Comment (dalippa) [inline papers/zhang_macdonald/zhang_macdonald.rst:53]: interfere => conflict

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 21]
```diff
@@ -0,0 +1,368 @@
...
+Reproducible Environments for Reproducible Results
+--------------------------------------------------
+
+.. class:: abstract
+
+   Trustworthy results require reproducibility, which must include the full computational
+   environment used to produce and analyze the data. Creating perfectly reproducible
+   environments is difficult because of fundamental challenges in the software packaging
+   domain, and understanding these problems is necessary to mitigate them. Focusing on the Python
+   ecosystem and a universal package manager called Pulp_, this paper explores various approaches
```
Comment (dalippa) [inline papers/zhang_macdonald/zhang_macdonald.rst:23]: "universal package manager" is incredibly important, especially if it's language agnostic. If that's truly the case, it belongs in the abstract.

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 22]
```diff
@@ -0,0 +1,368 @@
...
+
+Software Packaging Basics
+=========================
+
+Software packaging and distribution is an inherently complex problem domain because software is
+dependent on other software. Ensuring that code can be installed with compatible requirements is a
+fundamental problem in the packaging domain. The needs of users, developers, and administrators
+frequently interfere, leading to increasingly complex solutions. Various communities (often formed
+around a single programming language) address the problem with their own tool set and workflows, each
+optimizing for a different balance of needs. Rapid evolution within these communities can increase
```
Comment (dalippa) [inline papers/zhang_macdonald/zhang_macdonald.rst:55]: "Rapid evolution within these communities can increase the already significant learning curve" - how so? Evolution of the language? Java has been the same for about 10-15 years; Python for the the past 8. This isn't "rapid". Please clarify this thought some more.

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 23]
```diff
@@ -0,0 +1,368 @@
...
+offer an additional component to address this problem; curated repositories. Curated repositories
+can be thought of as a hybrid approach; the community as a whole maintains the interoperability of the
+software in the repository, relieving some of the user responsibility. However, this
+option also has problems; it can be less flexible because projects need to use versions of dependencies
+that are available in specific repositories. This can be particularly troublesome for developers
+when different repositories have different major versions of software (see semver_). Because
+code is run with different dependency versions in different environments, reproducibility is hard
+to guarantee, particularly across platforms.
+
+Python Packaging Ecosystem
```
Comment (dalippa) [inline papers/zhang_macdonald/zhang_macdonald.rst:87]: What about Anaconda? It handles cases when dependencies rely on native libraries, something that is ignored in this paper, and is increasingly more important as Python uses more native code for the sake of speed.

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 24]
```diff
@@ -0,0 +1,368 @@
...
+Broadly, there have been two high level approaches to packaging, and comparing them is useful to
+demonstrate the challenges to improving environmental reproducibility. The simplest
+approach, (which will be referred to as "monolithic"), is to bundle code with all dependencies and
+distribute it together as a single software package. This pattern is most common for user-centric
+platforms like Windows and OSX. Monolithic packages are simple to install, but can sacrifice
+transparency and flexibility for initial user convenience. The administrators and the developers
+become responsible for maintenance of the whole environment. For cross-platform applications,
+packages may need to be built for every combination of hardware, operating system, and language
+version. This process can be difficult and time consuming, leading to long development cycles and
+infrequent updates. This pattern is leads to chronically outdated software, and when security is an
```
Comment (dalippa) [inline papers/zhang_macdonald/zhang_macdonald.rst:68]: Why is security of such a concern for researchers? Researchers often operate in sandboxed, not public, environments. If the goal of the paper was reproducibility in production environments, this argument has more weight. For research, not as much.

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 25]
```diff
@@ -0,0 +1,368 @@
...
+==========================
+
+Python has a strong community that facilitates packaging with many open source tools. The Python
+Package Index (PyPI) is a community repository that allows package authors to contribute their
+work and users to find and install whatever packages they need. Currently, the standard tool to
+install Python software from PyPI is pip_, which also installs any dependencies. PyPI is not a
+curated repository like Fedora, so dependency problems can be frequent. To address this, it is
+generally recommended by the community to use virtual environments [0]_. A "virtualenv" is used to
+to isolate installed Python projects (along with their dependencies), which allows the user to
+install software which might otherwise have conflicts. Additionally, each virtualenv can use a
```
Comment (dalippa) [inline papers/zhang_macdonald/zhang_macdonald.rst:97]: This needs a bit more clarity.

Actually, a virtualenv is a private Python repository so that the system environment doesn't pollute the local environment. It *also* helps with conflicting dependencies between local and system, but does *not* help with conflicts between dependencies within the scope of the application. That must be managed/mitigated by the developer.

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 26]
```diff
@@ -0,0 +1,368 @@
...
+compatibility.
+
+During development, keeping dependencies flexible is ideal because the latest versions of a
+dependency will have a longer life before they become obsolete. However, after development is
+complete and final analysis begins, the need for reproducibility overtakes the benefits of
+keeping dependencies up to date. Semantic versioning is considered a best practice in the Python
+community, and uses X.Y.Z version numbers. Even for backwards compatible release (Y or Z releases),
+a researcher would not want to use numpy-1.13.1 for part of their analysis and numpy-1.14.2 for
+another. The risk that subtle changes could affect the results is too great, so it is recommended
+that researchers pin their versions.
```
Comment (dalippa) [inline papers/zhang_macdonald/zhang_macdonald.rst:115]: When releasing *any* software, pinning versions is important.

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 27]
```diff
@@ -0,0 +1,368 @@
...
+
+A Basic Workflow with Python and Git
+------------------------------------
+
+Scholarly research containing descriptions of methodology is no longer sufficient. For standalone
+scripts, publishing source code might be acceptable, but as computational systems grow more
+complex, this method becomes more unreliable. Nontrivial research often depends on other external
+libraries for everything from left-padding a line to a framework for scalable machine learning.
+This "has led to an ever larger and more complex black box between what was actually done and what
+is described in literature." :cite:`Boettiger`
```
Comment (dalippa) [inline papers/zhang_macdonald/zhang_macdonald.rst:126]: Have you compared your method to something like Vagrant or Chef where you can create an entire *system* environment, which would be less flexible (e.g. for cloud computing) but provides a true private sandbox?

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 28]
```diff
@@ -0,0 +1,368 @@
...
+
+Using an online git repository is a great way to keep track of source code :cite:`Wilson`.  With
+git you can easily track changes to data and software. Git identifies commits by a unique hash,
+which can be used to reference a specific point in the source code, practically guaranteeing the
+correct bits. This approach is incomplete however, because git lacks the ability to do
```
Comment (dalippa) [inline papers/zhang_macdonald/zhang_macdonald.rst:131]: What about submodules (https://git-scm.com/book/en/v2/Git-Tools-Submodules)?

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 29]
```diff
@@ -0,0 +1,368 @@
...
+
+Each of the approaches discussed offer a fundamental tradeoff when choosing a package management
+strategy. Strategies that increase control can improve reliability, but put significantly more
+responsibility on the maintainers. Even if a particular strategy well works for a specific project
+in its ecosystem, another ecosystem with a different tool set may not fit the strategy the same
+way, and will also come with a new learning curve.
+
+An alternative to the eclectic strategies native to various ecosystems is a universal package
+manager like Pulp [2]_. Pulp is a fully open source Python project that manages packages of any type by
+leveraging a plugin architecture. With the python plugin, for example, Pulp is able to
```
Comment (dalippa) [inline papers/zhang_macdonald/zhang_macdonald.rst:164]: Are there plugins for other languages? Is this truly a universal package manager?

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 30]
```diff
@@ -0,0 +1,368 @@
...
+This section discusses workflows at a very high level and does not include all steps for brevity.
+The Pulp documentation should be referenced for comprehensive workflows and specific commands [4]_.
+
+A Pulp server can be set up and configured to fetch the latest dependencies from PyPI, and pip can be
+configured to install from a hosted Pulp publication. Each time Pulp fetches new content, it
+creates a new repository version. Development is never blocked because the administrator can
+instantly (without downtime) roll back to a stable version whenever there is a problem.
+
+When the project matures enough to be used in publishable research, a curated repository is created
+containing only the desired versions of packages. Source code should be packaged with twine,
```
Comment (dalippa) [inline papers/zhang_macdonald/zhang_macdonald.rst:239]: twine?

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 31]
```diff
@@ -0,0 +1,368 @@
...
+of a curated repository is shared, it can be used to create a Python environment with exactly the
+same bits; the procedure documentation of the research should include instructions for configuring
+pip to use this publication.
+
+Even as research proceeds through peer review and publication, development can still continue
+against recently updated dependencies. Each Pulp publication is isolated and immutable, allowing
+legacy publications that ensure reproducibility to be served parallel to new publications used for
+flexible, up-to-date development.
+
+If a security flaw is discovered in a dependency that was used in published research, a new
```
Comment (dalippa) [inline papers/zhang_macdonald/zhang_macdonald.rst:250]: What about deprecation in case there is a security flaw with the package itself?

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 32]
```diff
@@ -0,0 +1,368 @@
...
+Pulp does not manage content itself, but instead relies on plugins to add support for one
+content type or another. Research labs with custom package types can create a plugin for Pulp
+to manage those artifacts. A basic plugin can be created by defining the custom content type.
+Just by defining the basic content type, the lab gains the ability to upload content, and manage
+it through curated repositories. More complex features such as syncing from a remote source,
+dependency solving, and custom publications can be optionally added for a feature rich
+content management platform. Plugin development is well documented [5]_, and the pulp_plugin package
+is semantically versioned.
+
+Summary
```
Comment (dalippa) [inline papers/zhang_macdonald/zhang_macdonald.rst:312]: Future work section?

File: papers/zhang_macdonald/zhang_macdonald.rst [hunk 33]
```diff
@@ -0,0 +1,368 @@
...
+================
+
+We appreciate Red Hat's continued support for open source technologies
+(including Pulp), and to the PyPA for their continuous effort at making
+Python packaging usable and stable. A special thank you to Michael Hrivnak,
+who helped formulate and fact check the Containers terminology, and Dana
+Walker for proof reading.
+
+
+References
```
Comment (dalippa) [inline papers/zhang_macdonald/zhang_macdonald.rst:331]: Please use APA style for references.

**Review by dalippa — CHANGES_REQUESTED** [review]
@werwty @WorklnProgress @proceedings Overall the paper is improved, but I recommend we reject it on the basis that there are claims that are not backed up or are incorrect, alternatives that are not explored, and issues in the narrative & citations.

I am open to the possibility of an extension, so that the paper would be published after the conference with the final copy of the proceedings.

**Comment by deniederhut** [comment]
Hi @werwty ! I know this is a busy time with the conference in full swing, but will you have time to respond to these comments?

**Comment by asmacdo** [comment]
@deniederhut, although we'd love to clean it up and get this into the proceedings, I don't think we will have time during the conference. Feel free to close. Thanks for all the review, maybe we can get this in next year :)

**Comment by deniederhut** [comment]
I'm sorry to hear this, but thanks for your honesty. Closing as withdrawan.
